kvserver: don't subsume self-blocking latency in scheduler latency

tbg opened the pull request:
We noticed elevated p99s during scale testing that disappeared when disabling
internal timeseries storage. The theory is that the timeseries replicas are
quite busy, and the scheduler latency was (unintentionally, I argue) picking up
raft **handling** latencies because it wasn't accounting for the case in which
handling a replica was delayed by an inflight handling of that same replica.

This commit addresses this issue, so that the scheduling latency metrics truly
reflect only the scheduling latency.

Fixes #147911
Informs #147758

Release note: None

tbg committed:
kvserver: don't subsume self-blocking latency in scheduler latency

We noticed elevated p99s during scale testing that disappeared when disabling
internal timeseries storage. The theory is that the timeseries replicas are
quite busy, and the scheduler latency was (unintentionally, I argue) picking up
raft **handling** latencies because it wasn't accounting for the case in which
handling a replica was delayed by an inflight handling of that same replica.

This commit addresses this issue, so that the scheduling latency metrics truly
reflect only the scheduling latency.

Fixes #147911

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147913)
<!-- Reviewable:end -->

tbg commented:
Hmm, having trouble showing the benefit of this change. I thought that if I disabled range splits and ran a kv0 workload against the range, alongside a change that adds a floor of 1ms for processing anything on a given range, I would prominently see the self-blocking, and the scheduler latency p50 (or at least some of the higher quantiles) should end up in the >1ms range and this should disappear with the fix commit. But with the setup below, even the p99 is in the 100us range, both with and without the fix. I still think the effect should be there. But I don't know what's wrong with the experiment either.
Will put this on hold for now.

```bash
#!/bin/bash
set -euo pipefail

c=local
(roachprod create -n 1 $c || true) &> /dev/null
roachprod stop $c
./dev build
roachprod put $c cockroach
roachprod start $c:1
roachprod sql $c:1 -- -e 'set cluster setting kv.split_queue.enabled = false;'
roachprod run $c:1 -- ./cockroach workload run kv --init --read-percent 0 --concurrency 10 --max-rate 20 --duration 100s {pgurl:1}

```

<details><summary>Patch</summary>
<p>

```diff
From 1ddc8c8f879b852e31a0f59a1333ce276f707b3c Mon Sep 17 00:00:00 2001
From: Tobias Grieger <tobias.b.grieger@gmail.com>
Date: Tue, 10 Jun 2025 18:54:25 +0200
Subject: [PATCH] [dnm] 1ms sleep

---
 pkg/kv/kvserver/replica_raft.go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pkg/kv/kvserver/replica_raft.go b/pkg/kv/kvserver/replica_raft.go
index 600a15ce34ba..1004f2fe8da5 100644
--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -855,7 +855,6 @@ var noSnap IncomingSnapshot
 func (r *Replica) handleRaftReady(
 	ctx context.Context, inSnap IncomingSnapshot,
 ) (handleRaftReadyStats, error) {
-
 	// Don't process anything if this fn returns false.
 	if fn := r.store.cfg.TestingKnobs.DisableProcessRaft; fn != nil && fn(r.store.StoreID()) {
 		return handleRaftReadyStats{
@@ -889,6 +888,7 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 	stats = handleRaftReadyStats{
 		tBegin: crtime.NowMono(),
 	}
+	time.Sleep(time.Millisecond)
 	defer func() {
 		stats.tEnd = crtime.NowMono()
 	}()

```

</p>
</details>

tbg commented:
@pav-kv it's annoying that I wasn't able to demonstrate the benefit of the change. I still think semantically, this makes sense and would give us a more trustworthy. Let's discuss whether to merge this or not - maybe I missed something in how this all works.

pav-kv commented on pkg/kv/kvserver/scheduler.go:
```diff
@@ -453,6 +453,12 @@ func (ss *raftSchedulerShard) worker(
+			// can not possibly happen before the current processing is done (i.e.
+			// now). We do not want the scheduler latency to pick up the time spent
+			// handling this replica.
+			state.begin = crtime.NowMono()
```
Is this a no-op? Should we rather update `ss.state[id].begin`?

tbg replied:
No, this wasn't it - I had trouble even elevating the baseline latency on master. The bug here was real and is now fixed, though, and I added a unit test.

tbg committed:
kvserver: add scheduler re-add test

tbg commented:
@pav-kv PTAL. I added a proper unit test, so I don't have to figure out why I wasn't able to trigger the self-blocking in the end-to-end test setup.

tbg commented on pkg/kv/kvserver/scheduler_test.go:
```diff
@@ -352,6 +367,75 @@ func TestSchedulerBuffering(t *testing.T) {
+func TestSchedulerEnqueueWhileProcessing(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.UnderNonTestBuild(t) // stateTestIntercept needs crdb test build
```
I usually run tests without the build tag through the IDE, I assume others do as well since it takes extra work to enable the crdb_test tag.

tbg committed:
kvserver: drop an unused ctx

tbg committed:
kvserver: address nits

tbg committed:
kvserver: rename a field

This is a more evocative name - we're measuring when the event was queued, but
not when it was processed.

tbg committed:
kvserver: explain an invariant

tbg committed:
kvserver: harmonize assertion guard

There was an existing assertion guarded behind util.RaceEnabled; guard the new
assertions in the same way.

The test still relies on CrdbTestBuild only.

tbg reviewed (commented):
TFTR!

I addressed most of the comments. I agree that there are improvements left to make, but I'm going to stop adding to this now and move on to other things.

bors r+
